docs/features/incremental-generators.cookbook.md

```diff
@@ -0,0 +1,656 @@
+not accept `SyntaxNode`s. It instead requires getting the `string` representation and putting that into a `SourceText`. Instead of `SyntaxNode`, we recommend using a
+wrapper around `StringBuilder` that will keep track of indent level and prepend the right amount of indentation when `AppendLine` is called. See
+[this](https://github.com/dotnet/roslyn/issues/52914#issuecomment-1732680995) conversation on the performance of `NormalizeWhitespace` for more examples, performance
+measurements, and discussion on why we don't believe that `SyntaxNode`s are a good abstraction for this use case.
+
+## Designs
+
+This section is broken down by user scenarios, with general solutions listed first, and more specific examples later on.
+
+### Generated class
+
+**User scenario:** As a generator author I want to be able to add a type to the compilation, that can be referenced by the user's code. Common use cases include
+creating an attribute that will be used to drive other source generator steps.
+
+**Solution:** Have the user write the code as if the type was already present. Generate the missing type based on information available in the compilation using
+the `RegisterPostInitializationOutput` step.
+
+**Example:**
+
+Given the following user code:
+
+```csharp
+public partial class UserClass
+{
+    [GeneratedNamespace.GeneratedAttribute]
+    public partial void UserMethod();
+}
+```
+
+Create a generator that will create the missing type when run:
+
+```csharp
+[Generator]
+public class CustomGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        context.RegisterPostInitializationOutput(static postInitializationContext => {
+            postInitializationContext.AddSource("myGeneratedFile.cs", SourceText.From("""
+                using System;
+
+                namespace GeneratedNamespace
+                {
+                    internal sealed class GeneratedAttribute : Attribute
+                    {
+                    }
+                }
+                """, Encoding.UTF8));
+        });
+    }
+}
+```
+
+**Alternative Solution**: If you are also providing a library to your users, in addition to a source generator, simply have that library include the
+attribute definition.
+
+### Additional file transformation
+
+**User scenario:** As a generator author I want to be able to transform an external non-C# file into an equivalent C# representation.
+
+**Solution:** Use the `AdditionalTextsProvider` to filter for and retrieve your files. Transform them into the code you care about, then
+register that code with the solution.
+
+**Example:**
+
+```csharp
+[Generator]
+public class FileTransformGenerator : IIncrementalGenerator
+{
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        var pipeline = context.AdditionalTextsProvider.Select(static (text, cancellationToken) =>
```

Comment:
We should probably re-write this example as a `Where(EndsWith(".xml")).Select(...)` then you don't need the second `Where` to filter out the defaults.